### DISCUSSION TITLE
Plugins - Support extending Warp with custom plugins and extensions

### DISCUSSION THREAD

**Discussion by shankarshastri** [warpdotdev/Warp#435] [Product ideas]
### Describe the solution you'd like?

```shell
I'm not sure whether it makes sense, but still just a small proposal !!

Nowadays we use code editors such as VSCode and others, which supports adding in plugins/extensions for a different set of use cases. I understand that Warp already is a terminal, and you can do whatever you want using shell commands.

But will it be possible for the community to contribute back for a Warp extension may be a theme or any feature extension?
```

### Is your feature request related to a problem? Please describe.

```shell
I'm not sure whether it makes sense, but still just a small proposal !!
```

**Comment by gagata** [comment]
Thanks for creating this! We're working towards making Warp more extensible in the future. First thing that's on our plate is allowing for custom themes so stay tuned! :) 

Is there anything you have in mind (extension/plugin) that you truly miss in Warp nowadays?

**Comment by shankarshastri** [comment]
From a developer perspective few examples: a rest-client, a websocket client !!

**Comment by elviskahoro** [comment]
Please comment with plugin ideas as well, particularly existing plugins that could take advantage and build on some of Warp's distinguishing features. Particularly:

1. Blocks
2. Input Editor
3. Cloud Features
4. [Underlying Data Model](https://blog.warp.dev/how-warp-works/) - See Data Model subsection under Implementing Blocks
5. UI/UX

**Reply by Alexndrrr** [reply]
I think bringing some of these features to more apps than just the shell itself would be awesome. E.g. support for code blocks and full input editor/autocomplete features within an iPython REPL.

**Reply by elviskahoro** [reply]
Ah! Yeah that's the plan!
https://github.com/warpdotdev/Warp/discussions/641

**Comment by elviskahoro** [comment]
FZF+Ranger!

> Zsh keybindings. 

> https://github.com/junegunn/fzf/blob/master/shell/completion.
> zsh I often use Ctrl-space to open fzf to search for a file nested in some directory. When I want to cd into a dir, I usually am thinking about the final directory name, and not the full path to get there. It would be fantastic to open up a fuzzy find window to search for files and directories and have the result inserted into the command line.

**Reply by elviskahoro** [reply]
Need to think some more on creating a FZF API so people can leverage our built-in version for improved performance and to avoid extra installation steps.

**Comment by elviskahoro** [comment]
From epher via Discord - https://discord.com/channels/851854972600451112/906269689116770304/906350632263376916

> This is pretty vanilla maybe but I’d love to have extensions to the terminal that don’t need to be written as a bunch of shell scripts that ultimately manipulate environment variables. Case in point would be something like a git (status etc.) integration. Why not write the logic in rust and integrate it into the actual UI rather than into the command prompt. I guess I’m thinking something like a sublime text / vscode plugin ecosystem. Maybe for ease of use (also to support the browser) the plugins could run as js/wasm (but compilable from rust with a proper rust api if the plugin devs are game)

**Reply by elviskahoro** [reply]
From CA_Lobo via Discord - https://discord.com/channels/851854972600451112/906350632263376916/906393591188963368
> I'd expand ephemer's idea to a library of preconfigured snippets that can be selected individuallly or chained/piped together... a command palette selection + pipe cmd + a second command palette selection or an exisiting snippet... add a textual UI user query to this mix cuz I'm always forgetting the right options for a script that I use maybe twice a month....  and perhaps a oneline description for each snippet as a reminder... user case is that most the stuff I do in the terminal window is repetive and needs to be automated...  and avoid typing rm -rf *  when I meant rm -rf *.build ...

**Reply by elviskahoro** [reply]
From CA_Lobo via Discord
> Excellent start! I like it! Glad to see others have the same ideas! Do we have a UI to go along with it?  Can the UI chain existing tasks together in the same principles of a Unix tools- eg multiple reusable individual tasks when chained together can automate a multi step process? something along the lines of "do a clean development checkout, do a debug build, package it, smoke test it and notify me and the QA team if passes, and any failures notify me" I can't wait to play with this as I have a dozen "scraps of paper" receipes on my computer to deal with repetitive tasks... humm, in a previous life, one thing I'd love to be able to do would be to share these tasks/receipes with others... so exporting/importing of packaged tasks would be something to think about...

**Reply by elviskahoro** [reply]
See the Workflows V1 memo!
https://github.com/warpdotdev/Warp/discussions/625

**Comment by elviskahoro** [comment]
From ShrubtheBub via Discord: https://discord.com/channels/851854972600451112/906269689116770304/928064929469173783

> Git conflict tui 
> Basically you cat a file If there is a git conflict, it helps you resolve  it

**Comment by elviskahoro** [comment]
> Ctrl-R already does the history search but fzf has Ctrl+T which also lets you fuzzy-search inside a folder to find files in a very deep hierarchy

**Comment by elviskahoro** [comment]
Built in functionality for jumping and navigating directories similarly to Zsh-Z.

I think we could overlap this with FZF, Ranger, and Context Chips.

**Reply by elviskahoro** [reply]
> most users do not want to use the mouse, i don't, thing i like about Z is that i type "z foo" and it will CD to the path containing "foo" that i use more, even the one containing "fou" if it has a very high score. It won't ask anything, so maybe i see this more in the Shell completion field or in a chip that changes while I type if i can then quickly use it via a handy shortcut like shift+enter

**Comment by elviskahoro** [comment]
> https://github.com/nvbn/thefuck

**Comment by ym-han** [comment]
Hi,

It would be really awesome if you could expose an API that would allow third-party devs to programmatically do things like:
- drawing / overlaying small graphics over characters 
- updating text in buffers, and ideally selecting and changing specific spans of text
- getting event callbacks when text in the buffer changes
- selecting arbitrary blocks

This would allow third-party devs to implement something like Cursorless (https://www.youtube.com/watch?v=JxcNW0hnfTk) for Warp (or, more realistically, implement a plugin that talks to a Cursorless server). Cursorless basically is a framework, built on top of the Talon Voice ecosystem, that lets users code by voice with Treesitter-like structural editing. Given that Warp basically has an editor within the terminal, it would be really awesome if there were an API that allowed us to build a plugin to get the sort of functionality for Warp that we currently have for editors like VSCode with Cursorless.

**Comment by dacort** [comment]
I would love the ability to programmatically create screenshots of blocks. 

I have two use-cases:
- documentation and demo building 
- experimenting with a new technology.

For the demo use-case, I often end up using screenshots or recordings in my demos or docs and I currently have to do that all manually.

For the experimentation use case, it's like a journal. I want to keep track of the commands I executed (and ideally share the whole set of blocks with somebody) so I can see how I got to a certain state. Almost like a runbook...

**Comment by felipetesc** [comment]
Would be doable or even possible to integrate(embed) something like Lapce into Warp ?

**Comment by hugows** [comment]
Please - for example I'd love to have a fullscreen message or a Pushbullet notification for (say) every terminal command that:
 1. takes longer than 30 seconds
 2. Warp is not focused 
 3. Current PWD matches some regex 

The current implementation generates a subtle notification but I frequently miss it.

This is the kind of stuff that would be so cool - and would keep differentiating from the other "very fast terminals" which don't ship those cooler, user facing features.

**Reply by elviskahoro** [reply]
I could see this falling under a custom trigger framework similar to iTerm's custom trigger function
